float32 x_ul
float32 y_ul
float32 z_ul
float32 x_lr
float32 y_lr
float32 z_lr
